{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ff6f0135",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "import scipy.stats as stats\n",
    "import math\n",
    "import re\n",
    "import subprocess\n",
    "from os import listdir\n",
    "from os.path import isfile, join"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "267dabf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl \n",
    "mpl.use('Agg')\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "plt.close(\"all\")\n",
    "mpl.rcParams['font.sans-serif'] = \"Arial\"\n",
    "mpl.rcParams[\"font.family\"] = 'sans-serif'\n",
    "mpl.rcParams['font.size'] = 8\n",
    "sns.set(style=\"whitegrid\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fa326275",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/davidy/jamboree20crispr/ontarget_analysis/230707_WTC11_Analysis\n"
     ]
    }
   ],
   "source": [
    "cd /Users/davidy/jamboree20crispr/ontarget_analysis/230707_WTC11_Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "5c833769",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrPerturbationTarget</th>\n",
       "      <th>startPerturbationTarget</th>\n",
       "      <th>endPerturbationTarget</th>\n",
       "      <th>name</th>\n",
       "      <th>strandPerturbationTarget</th>\n",
       "      <th>PerturbationTargetID</th>\n",
       "      <th>chrTSS</th>\n",
       "      <th>startTSS</th>\n",
       "      <th>endTSS</th>\n",
       "      <th>strandGene</th>\n",
       "      <th>...</th>\n",
       "      <th>avg_SeqCounts_Trt</th>\n",
       "      <th>log2fc_avg</th>\n",
       "      <th>log2fc_rep1</th>\n",
       "      <th>log2fc_rep2</th>\n",
       "      <th>z_log2fc_avg</th>\n",
       "      <th>z_log2fc_rep1</th>\n",
       "      <th>z_log2fc_rep2</th>\n",
       "      <th>z_log2fc_repavg</th>\n",
       "      <th>onebase_start</th>\n",
       "      <th>onebase_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>46464589</td>\n",
       "      <td>46464591</td>\n",
       "      <td>NA|chr2:46464589-46464591:+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr2:46464589-46464591:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>978.0</td>\n",
       "      <td>0.280608</td>\n",
       "      <td>0.150232</td>\n",
       "      <td>0.374436</td>\n",
       "      <td>0.981268</td>\n",
       "      <td>0.710833</td>\n",
       "      <td>1.117771</td>\n",
       "      <td>0.914302</td>\n",
       "      <td>46464588</td>\n",
       "      <td>46464589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>46464591</td>\n",
       "      <td>46464593</td>\n",
       "      <td>NA|chr2:46464591-46464593:+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr2:46464591-46464593:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>121.0</td>\n",
       "      <td>-0.643184</td>\n",
       "      <td>-0.638981</td>\n",
       "      <td>-0.646961</td>\n",
       "      <td>-0.422204</td>\n",
       "      <td>-0.475305</td>\n",
       "      <td>-0.306564</td>\n",
       "      <td>-0.390935</td>\n",
       "      <td>46464590</td>\n",
       "      <td>46464591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>46464569</td>\n",
       "      <td>46464571</td>\n",
       "      <td>NA|chr2:46464569-46464571:-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr2:46464569-46464571:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1739.0</td>\n",
       "      <td>1.006482</td>\n",
       "      <td>0.314522</td>\n",
       "      <td>1.562268</td>\n",
       "      <td>2.084051</td>\n",
       "      <td>0.957749</td>\n",
       "      <td>2.774197</td>\n",
       "      <td>1.865973</td>\n",
       "      <td>46464571</td>\n",
       "      <td>46464572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>46464607</td>\n",
       "      <td>46464609</td>\n",
       "      <td>NA|chr2:46464607-46464609:-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr2:46464607-46464609:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>29.0</td>\n",
       "      <td>-0.657438</td>\n",
       "      <td>-0.496606</td>\n",
       "      <td>-0.796025</td>\n",
       "      <td>-0.443860</td>\n",
       "      <td>-0.261324</td>\n",
       "      <td>-0.514433</td>\n",
       "      <td>-0.387879</td>\n",
       "      <td>46464609</td>\n",
       "      <td>46464610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>46464648</td>\n",
       "      <td>46464650</td>\n",
       "      <td>NA|chr2:46464648-46464650:+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr2:46464648-46464650:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>207.5</td>\n",
       "      <td>-0.907584</td>\n",
       "      <td>-0.628840</td>\n",
       "      <td>-1.201189</td>\n",
       "      <td>-0.823893</td>\n",
       "      <td>-0.460064</td>\n",
       "      <td>-1.079434</td>\n",
       "      <td>-0.769749</td>\n",
       "      <td>46464647</td>\n",
       "      <td>46464648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12697</th>\n",
       "      <td>chr7</td>\n",
       "      <td>6947558</td>\n",
       "      <td>6947560</td>\n",
       "      <td>NA|chr7:6947558-6947560:+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr7:6947558-6947560:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>88.0</td>\n",
       "      <td>-0.672463</td>\n",
       "      <td>-0.670587</td>\n",
       "      <td>-0.670541</td>\n",
       "      <td>-0.466685</td>\n",
       "      <td>-0.522807</td>\n",
       "      <td>-0.339446</td>\n",
       "      <td>-0.431126</td>\n",
       "      <td>6947557</td>\n",
       "      <td>6947558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12698</th>\n",
       "      <td>chr7</td>\n",
       "      <td>6964585</td>\n",
       "      <td>6964587</td>\n",
       "      <td>NA|chr7:6964585-6964587:-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7:6964585-6964587:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>482.5</td>\n",
       "      <td>-0.291004</td>\n",
       "      <td>-0.300932</td>\n",
       "      <td>-0.280858</td>\n",
       "      <td>0.112847</td>\n",
       "      <td>0.032761</td>\n",
       "      <td>0.203966</td>\n",
       "      <td>0.118363</td>\n",
       "      <td>6964587</td>\n",
       "      <td>6964588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12699</th>\n",
       "      <td>chr7</td>\n",
       "      <td>6978813</td>\n",
       "      <td>6978815</td>\n",
       "      <td>NA|chr7:6978813-6978815:+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr7:6978813-6978815:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>350.5</td>\n",
       "      <td>-0.409383</td>\n",
       "      <td>-0.572145</td>\n",
       "      <td>-0.253189</td>\n",
       "      <td>-0.067001</td>\n",
       "      <td>-0.374855</td>\n",
       "      <td>0.242549</td>\n",
       "      <td>-0.066153</td>\n",
       "      <td>6978812</td>\n",
       "      <td>6978813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12700</th>\n",
       "      <td>chr7</td>\n",
       "      <td>6993727</td>\n",
       "      <td>6993729</td>\n",
       "      <td>NA|chr7:6993727-6993729:+</td>\n",
       "      <td>+</td>\n",
       "      <td>chr7:6993727-6993729:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>44.5</td>\n",
       "      <td>-0.629059</td>\n",
       "      <td>-0.788572</td>\n",
       "      <td>-0.460804</td>\n",
       "      <td>-0.400744</td>\n",
       "      <td>-0.700130</td>\n",
       "      <td>-0.046969</td>\n",
       "      <td>-0.373550</td>\n",
       "      <td>6993726</td>\n",
       "      <td>6993727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12701</th>\n",
       "      <td>chr7</td>\n",
       "      <td>7009364</td>\n",
       "      <td>7009366</td>\n",
       "      <td>NA|chr7:7009364-7009366:-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7:7009364-7009366:-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>-0.464946</td>\n",
       "      <td>-0.419880</td>\n",
       "      <td>-0.486536</td>\n",
       "      <td>-0.151415</td>\n",
       "      <td>-0.146010</td>\n",
       "      <td>-0.082852</td>\n",
       "      <td>-0.114431</td>\n",
       "      <td>7009366</td>\n",
       "      <td>7009367</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12702 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrPerturbationTarget  startPerturbationTarget  endPerturbationTarget  \\\n",
       "0                      chr2                 46464589               46464591   \n",
       "1                      chr2                 46464591               46464593   \n",
       "2                      chr2                 46464569               46464571   \n",
       "3                      chr2                 46464607               46464609   \n",
       "4                      chr2                 46464648               46464650   \n",
       "...                     ...                      ...                    ...   \n",
       "12697                  chr7                  6947558                6947560   \n",
       "12698                  chr7                  6964585                6964587   \n",
       "12699                  chr7                  6978813                6978815   \n",
       "12700                  chr7                  6993727                6993729   \n",
       "12701                  chr7                  7009364                7009366   \n",
       "\n",
       "                              name strandPerturbationTarget  \\\n",
       "0      NA|chr2:46464589-46464591:+                        +   \n",
       "1      NA|chr2:46464591-46464593:+                        +   \n",
       "2      NA|chr2:46464569-46464571:-                        -   \n",
       "3      NA|chr2:46464607-46464609:-                        -   \n",
       "4      NA|chr2:46464648-46464650:+                        +   \n",
       "...                            ...                      ...   \n",
       "12697    NA|chr7:6947558-6947560:+                        +   \n",
       "12698    NA|chr7:6964585-6964587:-                        -   \n",
       "12699    NA|chr7:6978813-6978815:+                        +   \n",
       "12700    NA|chr7:6993727-6993729:+                        +   \n",
       "12701    NA|chr7:7009364-7009366:-                        -   \n",
       "\n",
       "           PerturbationTargetID  chrTSS  startTSS  endTSS  strandGene  ...  \\\n",
       "0      chr2:46464589-46464591:+     NaN       NaN     NaN         NaN  ...   \n",
       "1      chr2:46464591-46464593:+     NaN       NaN     NaN         NaN  ...   \n",
       "2      chr2:46464569-46464571:-     NaN       NaN     NaN         NaN  ...   \n",
       "3      chr2:46464607-46464609:-     NaN       NaN     NaN         NaN  ...   \n",
       "4      chr2:46464648-46464650:+     NaN       NaN     NaN         NaN  ...   \n",
       "...                         ...     ...       ...     ...         ...  ...   \n",
       "12697    chr7:6947558-6947560:+     NaN       NaN     NaN         NaN  ...   \n",
       "12698    chr7:6964585-6964587:-     NaN       NaN     NaN         NaN  ...   \n",
       "12699    chr7:6978813-6978815:+     NaN       NaN     NaN         NaN  ...   \n",
       "12700    chr7:6993727-6993729:+     NaN       NaN     NaN         NaN  ...   \n",
       "12701    chr7:7009364-7009366:-     NaN       NaN     NaN         NaN  ...   \n",
       "\n",
       "       avg_SeqCounts_Trt  log2fc_avg log2fc_rep1 log2fc_rep2 z_log2fc_avg  \\\n",
       "0                  978.0    0.280608    0.150232    0.374436     0.981268   \n",
       "1                  121.0   -0.643184   -0.638981   -0.646961    -0.422204   \n",
       "2                 1739.0    1.006482    0.314522    1.562268     2.084051   \n",
       "3                   29.0   -0.657438   -0.496606   -0.796025    -0.443860   \n",
       "4                  207.5   -0.907584   -0.628840   -1.201189    -0.823893   \n",
       "...                  ...         ...         ...         ...          ...   \n",
       "12697               88.0   -0.672463   -0.670587   -0.670541    -0.466685   \n",
       "12698              482.5   -0.291004   -0.300932   -0.280858     0.112847   \n",
       "12699              350.5   -0.409383   -0.572145   -0.253189    -0.067001   \n",
       "12700               44.5   -0.629059   -0.788572   -0.460804    -0.400744   \n",
       "12701               14.0   -0.464946   -0.419880   -0.486536    -0.151415   \n",
       "\n",
       "      z_log2fc_rep1  z_log2fc_rep2  z_log2fc_repavg  onebase_start  \\\n",
       "0          0.710833       1.117771         0.914302       46464588   \n",
       "1         -0.475305      -0.306564        -0.390935       46464590   \n",
       "2          0.957749       2.774197         1.865973       46464571   \n",
       "3         -0.261324      -0.514433        -0.387879       46464609   \n",
       "4         -0.460064      -1.079434        -0.769749       46464647   \n",
       "...             ...            ...              ...            ...   \n",
       "12697     -0.522807      -0.339446        -0.431126        6947557   \n",
       "12698      0.032761       0.203966         0.118363        6964587   \n",
       "12699     -0.374855       0.242549        -0.066153        6978812   \n",
       "12700     -0.700130      -0.046969        -0.373550        6993726   \n",
       "12701     -0.146010      -0.082852        -0.114431        7009366   \n",
       "\n",
       "       onebase_end  \n",
       "0         46464589  \n",
       "1         46464591  \n",
       "2         46464572  \n",
       "3         46464610  \n",
       "4         46464648  \n",
       "...            ...  \n",
       "12697      6947558  \n",
       "12698      6964588  \n",
       "12699      6978813  \n",
       "12700      6993727  \n",
       "12701      7009367  \n",
       "\n",
       "[12702 rows x 31 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'\\n\\n    \\n    h1 = sns.histplot(data=gq_rep_df_merged, x=\\'SeqCounts_Pla\\')\\n    plt.savefig(\\'WTC11_CRISPRi_Plasmid_SeqCounts.pdf\\')\\n    plt.close(\\'all\\')\\n    \\n    seqcounts = \\'SeqCounts_x\\'\\n    h2 = sns.histplot(data=gq_rep_df_merged, x=seqcounts)\\n    ofile = \\'WTC11_CRISPRi_\\' + str(rep) + \\'_\\' + seqcounts + \\'.pdf\\'\\n    plt.xscale(\\'log\\')\\n    plt.savefig(ofile)\\n    plt.close(\\'all\\')\\n    \\n    seqcounts = \\'SeqCounts_y\\'\\n    h2 = sns.histplot(data=gq_rep_df_merged, x=seqcounts)\\n    ofile = \\'WTC11_CRISPRi_\\' + str(rep) + \\'_\\' + seqcounts + \\'.pdf\\'\\n    plt.xscale(\\'log\\')\\n    plt.savefig(ofile)\\n    plt.close(\\'all\\')\\n    \\n\\n    # Add list of average z-scores to dictionary. The following section is sloppy, it\\'s just for graphing\\n    # I can do this crude list-based approach since the rep dfs all share the same index\\n    all_avg_z_scores[rep] = gq_rep_df_merged[\\'avg_z_log2fc\\'].tolist()\\n    # Exclude NT guides\\n    typed_df = gq_rep_df_merged.merge(pd.read_csv(gq_directory+\\'/\\'+pla_gq_file[0], sep = \\'\\t\\', usecols = [\\'name\\',\\'guideType\\']))\\n    #display(typed_df)\\n    targeting_df = typed_df.loc[typed_df[\\'guideType\\']==\\'targeting\\']\\n    targeting_avg_z_scores[rep] = typed_df.loc[typed_df[\\'guideType\\']==\\'targeting\\'][\\'avg_z_log2fc\\'].tolist()\\n               \\n\\n\\n    ## Plot correlation of replicate average z scores for only targeting guides\\n    ## Note: The following code works \\n    f, ax = plt.subplots(figsize=(6, 6))\\n    ax.scatter(x=targeting_avg_z_scores[\\'Rep1\\'], y=targeting_avg_z_scores[\\'Rep2\\'])\\n    #plt.xlim(0,1)\\n    #plt.ylim(0,1)\\n    #ax.plot([0, 1], [0, 1], ls=\"--\", c=\".3\")\\n    ax.set_xlabel(\\'Avg Z-score X\\')\\n    ax.set_ylabel(\\'Avg z-score Y\\')\\n    plt.savefig(\\'WTC11_CRISPRi_AvgZScores_TargetingGuides_Replicates_eq31.pdf\\'.format(rep))\\n    plt.close(\"all\")\\n\\n\\n\\n'"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### I/O\n",
    "all_files = [f for f in listdir('drive-download-20230708T035752Z-001/crispri/guideQuant') if isfile(join('drive-download-20230708T035752Z-001/crispri/guideQuant', f))]\n",
    "gq_files = [f for f in all_files if '_EQ_' not in f and '_files' not in f]\n",
    "\n",
    "### Indicate tags for control and treatment files\n",
    "control_tag = 'Unt'\n",
    "treatment_tag = '240mg'\n",
    "\n",
    "\n",
    "##### Process guideQuant files\n",
    "gq_directory = 'drive-download-20230708T035752Z-001/crispri/guideQuant'\n",
    "\n",
    "pla_gq_file = [f for f in gq_files if 'Pla' in f]\n",
    "control_gq_files = [f for f in gq_files if control_tag in f]\n",
    "treatment_gq_files = [f for f in gq_files if treatment_tag in f]\n",
    "\n",
    "all_avg_z_scores = {}\n",
    "targeting_avg_z_scores = {}\n",
    "\n",
    "combo_rep_df = []\n",
    "\n",
    "### Make a combined rep df with both tech reps for both conditions\n",
    "## For control\n",
    "control_rep_df = pd.read_csv(gq_directory+'/'+control_gq_files[0], sep = '\\t', usecols=['name','SeqCounts']).rename(columns={'SeqCounts':'SeqCounts_Ctrl1'}).merge(pd.read_csv(gq_directory+'/'+control_gq_files[1], sep = '\\t', usecols=['name','SeqCounts']).rename(columns={'SeqCounts':'SeqCounts_Ctrl2'}), on='name')\n",
    "# Create an average count column\n",
    "control_rep_df['avg_SeqCounts_Ctrl'] = control_rep_df[['SeqCounts_Ctrl1','SeqCounts_Ctrl2']].mean(axis=1)\n",
    "## For treatment\n",
    "treatment_rep_df = pd.read_csv(gq_directory+'/'+treatment_gq_files[0], sep = '\\t', usecols=['name','SeqCounts']).rename(columns={'SeqCounts':'SeqCounts_Trt1'}).merge(pd.read_csv(gq_directory+'/'+treatment_gq_files[1], sep = '\\t', usecols=['name','SeqCounts']).rename(columns={'SeqCounts':'SeqCounts_Trt2'}), on='name')\n",
    "# Create an average count column\n",
    "treatment_rep_df['avg_SeqCounts_Trt'] = treatment_rep_df[['SeqCounts_Trt1','SeqCounts_Trt2']].mean(axis=1)\n",
    "\n",
    "## Merge \n",
    "gq_rep_df_merged = control_rep_df.merge(treatment_rep_df, on='name')\n",
    "\n",
    "## Apply a pseudocount to any element under sequencing count coverage, and get normalization info\n",
    "SeqCount_pseudocount = 10\n",
    "condition_normalization_dict = {} # Key = Colname; Value = {'mean':X, 'total_count':Y}\n",
    "for col in [col for col in gq_rep_df_merged.columns.values if col != 'name']:\n",
    "    # Apply pseudocount\n",
    "    gq_rep_df_merged[str(col)].clip(lower=SeqCount_pseudocount,inplace=True)\n",
    "    # Get mean and total counts\n",
    "    condition_normalization_dict[col] = {'mean':gq_rep_df_merged[str(col)].mean(),\n",
    "                                        'total_count':gq_rep_df_merged[str(col)].sum()}\n",
    "\n",
    "\n",
    "### Calculate log2FC according to eq 3.1\n",
    "log2FC_pseudocount = 1\n",
    "gq_rep_df_merged['log2fc_avg'] = np.log2((log2FC_pseudocount + gq_rep_df_merged['avg_SeqCounts_Trt']/float(condition_normalization_dict['avg_SeqCounts_Trt']['mean'])) / (log2FC_pseudocount + gq_rep_df_merged['avg_SeqCounts_Ctrl']/float(condition_normalization_dict['avg_SeqCounts_Ctrl']['mean'])))\n",
    "gq_rep_df_merged['log2fc_rep1'] = np.log2((log2FC_pseudocount + gq_rep_df_merged['SeqCounts_Trt1']/float(condition_normalization_dict['SeqCounts_Trt1']['mean'])) / (log2FC_pseudocount + gq_rep_df_merged['SeqCounts_Ctrl1']/float(condition_normalization_dict['SeqCounts_Ctrl1']['mean'])))\n",
    "gq_rep_df_merged['log2fc_rep2'] = np.log2((log2FC_pseudocount + gq_rep_df_merged['SeqCounts_Trt2']/float(condition_normalization_dict['SeqCounts_Trt2']['mean'])) / (log2FC_pseudocount + gq_rep_df_merged['SeqCounts_Ctrl2']/float(condition_normalization_dict['SeqCounts_Ctrl2']['mean'])))\n",
    "\n",
    "### z-transform log2FC\n",
    "## Get mean and stdev of all avg_log2FCs for all guides\n",
    "## Note: DO NOT CALCULATE AVG Z SCORE FROM AN AVG LOG2FC\n",
    "allgrna_log2fc_avg_mean = np.mean(gq_rep_df_merged['log2fc_avg'])\n",
    "allgrna_log2fc_avg_sd = np.std(gq_rep_df_merged['log2fc_avg'], ddof = 1)\n",
    "allgrna_log2fc_rep1_mean = np.mean(gq_rep_df_merged['log2fc_rep1'])\n",
    "allgrna_log2fc_rep1_sd = np.std(gq_rep_df_merged['log2fc_rep1'], ddof = 1)\n",
    "allgrna_log2fc_rep2_mean = np.mean(gq_rep_df_merged['log2fc_rep2'])\n",
    "allgrna_log2fc_rep2_sd = np.std(gq_rep_df_merged['log2fc_rep2'], ddof = 1)\n",
    "\n",
    "# z-transform\n",
    "gq_rep_df_merged['z_log2fc_avg'] = (gq_rep_df_merged['log2fc_avg'] - allgrna_log2fc_avg_mean) / allgrna_log2fc_avg_sd\n",
    "gq_rep_df_merged['z_log2fc_rep1'] = (gq_rep_df_merged['log2fc_rep1'] - allgrna_log2fc_rep1_mean) / allgrna_log2fc_rep1_sd\n",
    "gq_rep_df_merged['z_log2fc_rep2'] = (gq_rep_df_merged['log2fc_rep2'] - allgrna_log2fc_rep2_mean) / allgrna_log2fc_rep2_sd\n",
    "\n",
    "# Add average of replicates z-l2fc\n",
    "gq_rep_df_merged['z_log2fc_repavg'] = (gq_rep_df_merged['z_log2fc_rep1']+gq_rep_df_merged['z_log2fc_rep2'])/2\n",
    "\n",
    "### Add in meta-data\n",
    "combo_rep_df = pd.read_csv(gq_directory+'/'+pla_gq_file[0], sep = '\\t').drop(columns=['SeqCounts']).merge(gq_rep_df_merged, on='name')\n",
    "\n",
    "# Add single base chromosomal coordinate based on strand\n",
    "combo_rep_df['onebase_start'] = np.where(combo_rep_df['strandPerturbationTarget']== '+', combo_rep_df['startPerturbationTarget']-1, combo_rep_df['endPerturbationTarget'])\n",
    "combo_rep_df['onebase_end'] = np.where(combo_rep_df['strandPerturbationTarget']== '+', combo_rep_df['startPerturbationTarget'], combo_rep_df['endPerturbationTarget']+1)\n",
    "\n",
    "    \n",
    "display(combo_rep_df)\n",
    "\n",
    "##### Write to output\n",
    "# Write full df\n",
    "combo_rep_df.to_csv('WTC11_CRISPRi_{0}_{1}_ProcessedGuideQuant.tsv'.format(treatment_tag,control_tag), sep = '\\t', index=False)\n",
    "# Write simple bed (with PAM coordinates) for intersection\n",
    "combo_rep_df.to_csv('WTC11_CRISPRi_{0}_{1}_Processed_Simple_PamCC.bed'.format(treatment_tag,control_tag), sep = '\\t', index=False, header=False, columns=['chrPerturbationTarget','startPerturbationTarget','endPerturbationTarget','name','z_log2fc_repavg'])\n",
    "# Write simple bed (with onebase coordinates) for intersection\n",
    "combo_rep_df.to_csv('WTC11_CRISPRi_{0}_{1}_Processed_Simple_SingleBaseCC.bed'.format(treatment_tag,control_tag), sep = '\\t', index=False, header=False, columns=['chrPerturbationTarget','onebase_start','onebase_end','name','z_log2fc_repavg'])\n",
    "\n",
    "# Write simple qBed for plotting\n",
    "combo_rep_df.to_csv('WTC11_CRISPRi_{0}_{1}_Processed_Simple_SingleBaseCC.qBed'.format(treatment_tag,control_tag), sep = '\\t', index=False, header=False, columns=['chrPerturbationTarget','onebase_start','onebase_end','z_log2fc_repavg'])\n",
    "\n",
    "\n",
    "##### Plotting\n",
    "\n",
    "f, ax = plt.subplots(figsize=(6, 6))\n",
    "ax.scatter(x=combo_rep_df['SeqCounts_Trt1'], y=combo_rep_df['SeqCounts_Trt2'])\n",
    "ax.set_xlabel('Counts Trt1 {}'.format(treatment_tag))\n",
    "ax.set_ylabel('Counts Trt2 {}'.format(treatment_tag))\n",
    "plt.savefig('WTC11_CRISPRi_{}_GQ_replicate_counts.pdf'.format(treatment_tag))\n",
    "plt.close(\"all\")\n",
    "\n",
    "f, ax = plt.subplots(figsize=(6, 6))\n",
    "ax.scatter(x=combo_rep_df['SeqCounts_Ctrl1'], y=combo_rep_df['SeqCounts_Ctrl2'])\n",
    "ax.set_xlabel('Counts Ctrl1 {}'.format(control_tag))\n",
    "ax.set_ylabel('Counts Ctrl2 {}'.format(control_tag))\n",
    "plt.savefig('WTC11_CRISPRi_{}_GQ_replicate_counts.pdf'.format(control_tag))\n",
    "plt.close(\"all\")\n",
    "\n",
    "## Plot correlation of replicate average z scores\n",
    "## Note: The following code works \n",
    "f, ax = plt.subplots(figsize=(6, 6))\n",
    "ax.scatter(x=combo_rep_df['z_log2fc_rep1'], y=combo_rep_df['z_log2fc_rep2'])\n",
    "ax.set_xlabel('Z-score Rep1 {} vs {}'.format(treatment_tag,control_tag))\n",
    "ax.set_ylabel('Z-score Rep2 {} vs {}'.format(treatment_tag,control_tag))\n",
    "plt.savefig('WTC11_CRISPRi_ZScores_Replicates_eq31_{0}_{1}_.pdf'.format(treatment_tag,control_tag))\n",
    "plt.close(\"all\")\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "    \n",
    "    h1 = sns.histplot(data=gq_rep_df_merged, x='SeqCounts_Pla')\n",
    "    plt.savefig('WTC11_CRISPRi_Plasmid_SeqCounts.pdf')\n",
    "    plt.close('all')\n",
    "    \n",
    "    seqcounts = 'SeqCounts_x'\n",
    "    h2 = sns.histplot(data=gq_rep_df_merged, x=seqcounts)\n",
    "    ofile = 'WTC11_CRISPRi_' + str(rep) + '_' + seqcounts + '.pdf'\n",
    "    plt.xscale('log')\n",
    "    plt.savefig(ofile)\n",
    "    plt.close('all')\n",
    "    \n",
    "    seqcounts = 'SeqCounts_y'\n",
    "    h2 = sns.histplot(data=gq_rep_df_merged, x=seqcounts)\n",
    "    ofile = 'WTC11_CRISPRi_' + str(rep) + '_' + seqcounts + '.pdf'\n",
    "    plt.xscale('log')\n",
    "    plt.savefig(ofile)\n",
    "    plt.close('all')\n",
    "    \n",
    "\n",
    "    # Add list of average z-scores to dictionary. The following section is sloppy, it's just for graphing\n",
    "    # I can do this crude list-based approach since the rep dfs all share the same index\n",
    "    all_avg_z_scores[rep] = gq_rep_df_merged['avg_z_log2fc'].tolist()\n",
    "    # Exclude NT guides\n",
    "    typed_df = gq_rep_df_merged.merge(pd.read_csv(gq_directory+'/'+pla_gq_file[0], sep = '\\t', usecols = ['name','guideType']))\n",
    "    #display(typed_df)\n",
    "    targeting_df = typed_df.loc[typed_df['guideType']=='targeting']\n",
    "    targeting_avg_z_scores[rep] = typed_df.loc[typed_df['guideType']=='targeting']['avg_z_log2fc'].tolist()\n",
    "               \n",
    "\n",
    "\n",
    "    ## Plot correlation of replicate average z scores for only targeting guides\n",
    "    ## Note: The following code works \n",
    "    f, ax = plt.subplots(figsize=(6, 6))\n",
    "    ax.scatter(x=targeting_avg_z_scores['Rep1'], y=targeting_avg_z_scores['Rep2'])\n",
    "    #plt.xlim(0,1)\n",
    "    #plt.ylim(0,1)\n",
    "    #ax.plot([0, 1], [0, 1], ls=\"--\", c=\".3\")\n",
    "    ax.set_xlabel('Avg Z-score X')\n",
    "    ax.set_ylabel('Avg z-score Y')\n",
    "    plt.savefig('WTC11_CRISPRi_AvgZScores_TargetingGuides_Replicates_eq31.pdf'.format(rep))\n",
    "    plt.close(\"all\")\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "2290d0bc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "##### Fig 2A. Avg sgRNA effects at element H3K27ac and ATAC (as a proxy of DHS) peaks\n",
    "\"\"\"\n",
    "Get significant elements. \n",
    "61 sig elements in 240mg, 32 sig elements in 160mg\n",
    "\n",
    "Note -- the order of the intersections below is intentional, and matches what was previously done\n",
    "Note -- There are some differences in processing algorithms, since gRNAs at the same DHS arent being measured orthogonally multiple times (e.g. FADS1, 2, and 3)\n",
    "\n",
    "First intersect the significant elements with 1 kb expanded TSSs (using hg38_mart_protein_coding_tss1kb.bed), excluding those that intersect\n",
    "Second, intersect significant elements with DHS/ATAC\n",
    "- For now, can ignore multi-summit peaks... Just make sure that I remove any duplicate peaks (these are multi-summit)\n",
    "Third, intersect the non-TSS sig cCRE DHS peaks with H3K27ac. No slop/range expansion! Remove duplicates.\n",
    "Fourth, if an element has multiple DHS peaks that intersect H3K27ac, nominate a single DHS peak. Only one DHS peak per sig element!\n",
    "- Option 1: Use the DHS peak with the highest average (per base) signal.\n",
    "- Option 2: Empirically select the DHS peak closest to the top 3 sgRNAs\n",
    "Fifth, get the average effects of all sgRNAs intersecting a given epifeature peak.\n",
    "- Effect is the average z-transformed log2FC from both screen replicates\n",
    "- sgRNA intersections are based on PAM coordinates, with no further slop/range expansion\n",
    "- If a DHS intersects two H3K27ac peaks, treat each H3K27ac peak independently\n",
    "\"\"\"\n",
    "##### (0)\n",
    "### Process EQ files, get significant common elements, and write to bed\n",
    "eq_directory = 'drive-download-20230708T035752Z-001/crispri/elementQuant'\n",
    "elements_rep1 = pd.read_csv(eq_directory+'/160mg_ENCFF646RAN.tsv',sep='\\t',usecols=['chrom','chromStart','chromEnd','PerturbationTargetID'])\n",
    "\n",
    "# Write table for intersection purposes\n",
    "common_sig_elements_output = 'WTC11_CRISPRi_Sig_Elements_{}.bed'.format(treatment_tag)\n",
    "elements_rep1.to_csv(common_sig_elements_output, sep = '\\t', index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "a0e66543",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 40 common significant elements that intersect +/- 1kb expanded TSS ranges ['chr3:38024347-38025405:.', 'chr2:47034146-47034911:.', 'chr2:47368918-47370143:.', 'chr20:4221298-4222072:.', 'chr2:46616497-46617342:.', 'chr20:4686246-4687437:.', 'chr20:5911126-5911906:.', 'chrX:135031924-135033120:.', 'chr2:47175627-47177035:.', 'chr3:37860400-37861401:.', 'chrX:135342909-135344870:.', 'chr2:46698689-46699696:.', 'chr7:5513400-5514161:.', 'chr2:46541698-46544265:.', 'chr20:4148481-4149850:.', 'chr20:6005736-6006471:.', 'chr7:6058597-6059412:.', 'chr7:6374225-6375026:.', 'chr20:4249093-4249842:.', 'chr3:37175800-37176855:.', 'chrX:134373043-134373594:.', 'chr2:47570494-47571981:.', 'chr7:5562057-5563839:.', 'chr2:46940390-46941878:.', 'chrX:134796192-134797382:.', 'chr2:48314106-48315728:.', 'chr20:4172153-4172791:.', 'chrX:134914362-134915747:.', 'chr2:48440406-48441214:.', 'chr7:6104367-6105482:.', 'chrX:135098651-135099180:.', 'chr20:4822643-4824017:.', 'chr3:37242911-37243712:.', 'chr7:6483307-6484464:.', 'chr7:5592622-5594535:.', 'chr3:37998914-37999449:.', 'chr7:6348332-6349203:.', 'chr20:5610366-5611239:.', 'chr7:5529672-5532270:.', 'chr20:5112821-5113595:.']\n"
     ]
    }
   ],
   "source": [
    "##### (1) Intersect with 1 kb expanded TSSs, and exclude any overlapping elements\n",
    "# Note: Manually copied over hg38_mart_protein_coding_tss1kb.bed into 230707_WTC11_Analysis\n",
    "illegal_tss_ofile = 'illegal_tss_common_sig_elements_intersect.bed' \n",
    "fopen = open(illegal_tss_ofile, 'w')\n",
    "subprocess.call(['bedtools','intersect','-a',common_sig_elements_output,'-b','hg38_mart_protein_coding_tss1kb.bed','-wa','-wb'],stdout=fopen)\n",
    "fopen.close()\n",
    "\n",
    "### Get common sig non-TSS elements \n",
    "# Process illegal tss intersecting elements\n",
    "illegal_tss_common_elements = list(set(pd.read_csv(illegal_tss_ofile, sep = '\\t', header = None, names = ['chr1','start1','end1','PerturbTargetID','chr2','start2','end2','symbol'])['PerturbTargetID'].tolist()))\n",
    "print('There are {} common significant elements that intersect +/- 1kb expanded TSS ranges'.format(len(illegal_tss_common_elements)), illegal_tss_common_elements)\n",
    "# Exclude the illegal elements from the common sig elements\n",
    "non_promoter_common_sig_elements = common_sig_elements.loc[~common_sig_elements['PerturbationTargetID'].isin(illegal_tss_common_elements)]\n",
    "# Write this dataframe to a bed for intersection purposes\n",
    "tssfiltered_common_sig_elements_output = 'WTC11_CRISPRi_NonTSS_Common_Sig_Elements.bed'\n",
    "non_promoter_common_sig_elements.to_csv(tssfiltered_common_sig_elements_output, sep = '\\t', index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "8f0bbd2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1</th>\n",
       "      <th>start1</th>\n",
       "      <th>end1</th>\n",
       "      <th>PerturbTargetID</th>\n",
       "      <th>chr2</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "      <th>signal</th>\n",
       "      <th>dhs_cc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191521</td>\n",
       "      <td>47193363</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191548</td>\n",
       "      <td>47191727</td>\n",
       "      <td>1.017580</td>\n",
       "      <td>chr2_47191548_47191727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191521</td>\n",
       "      <td>47193363</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47192620</td>\n",
       "      <td>47192709</td>\n",
       "      <td>0.866826</td>\n",
       "      <td>chr2_47192620_47192709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr20</td>\n",
       "      <td>4760486</td>\n",
       "      <td>4761348</td>\n",
       "      <td>chr20:4760486-4761348:.</td>\n",
       "      <td>chr20</td>\n",
       "      <td>4760680</td>\n",
       "      <td>4760840</td>\n",
       "      <td>2.908270</td>\n",
       "      <td>chr20_4760680_4760840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr20</td>\n",
       "      <td>4760486</td>\n",
       "      <td>4761348</td>\n",
       "      <td>chr20:4760486-4761348:.</td>\n",
       "      <td>chr20</td>\n",
       "      <td>4760960</td>\n",
       "      <td>4761280</td>\n",
       "      <td>0.829138</td>\n",
       "      <td>chr20_4760960_4761280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5486761</td>\n",
       "      <td>5488068</td>\n",
       "      <td>chr7:5486761-5488068:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5486960</td>\n",
       "      <td>5487180</td>\n",
       "      <td>4.026350</td>\n",
       "      <td>chr7_5486960_5487180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5486761</td>\n",
       "      <td>5488068</td>\n",
       "      <td>chr7:5486761-5488068:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5487920</td>\n",
       "      <td>5488100</td>\n",
       "      <td>2.456010</td>\n",
       "      <td>chr7_5487920_5488100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5607061</td>\n",
       "      <td>5607714</td>\n",
       "      <td>chr7:5607061-5607714:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5607560</td>\n",
       "      <td>5607760</td>\n",
       "      <td>2.135660</td>\n",
       "      <td>chr7_5607560_5607760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr7</td>\n",
       "      <td>6525977</td>\n",
       "      <td>6526746</td>\n",
       "      <td>chr7:6525977-6526746:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>6526560</td>\n",
       "      <td>6526840</td>\n",
       "      <td>2.116820</td>\n",
       "      <td>chr7_6526560_6526840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5426774</td>\n",
       "      <td>5428241</td>\n",
       "      <td>chr7:5426774-5428241:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5426800</td>\n",
       "      <td>5426960</td>\n",
       "      <td>2.895700</td>\n",
       "      <td>chr7_5426800_5426960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5426774</td>\n",
       "      <td>5428241</td>\n",
       "      <td>chr7:5426774-5428241:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5427360</td>\n",
       "      <td>5427634</td>\n",
       "      <td>1.331650</td>\n",
       "      <td>chr7_5427360_5427634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5426774</td>\n",
       "      <td>5428241</td>\n",
       "      <td>chr7:5426774-5428241:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5427860</td>\n",
       "      <td>5428240</td>\n",
       "      <td>2.964800</td>\n",
       "      <td>chr7_5427860_5428240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr7</td>\n",
       "      <td>6663773</td>\n",
       "      <td>6664698</td>\n",
       "      <td>chr7:6663773-6664698:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>6663840</td>\n",
       "      <td>6664040</td>\n",
       "      <td>3.504990</td>\n",
       "      <td>chr7_6663840_6664040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr7</td>\n",
       "      <td>6081033</td>\n",
       "      <td>6081797</td>\n",
       "      <td>chr7:6081033-6081797:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>6081460</td>\n",
       "      <td>6081680</td>\n",
       "      <td>9.610470</td>\n",
       "      <td>chr7_6081460_6081680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5555056</td>\n",
       "      <td>5556791</td>\n",
       "      <td>chr7:5555056-5556791:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5554920</td>\n",
       "      <td>5555060</td>\n",
       "      <td>0.841701</td>\n",
       "      <td>chr7_5554920_5555060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5555056</td>\n",
       "      <td>5556791</td>\n",
       "      <td>chr7:5555056-5556791:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5555120</td>\n",
       "      <td>5555340</td>\n",
       "      <td>1.658280</td>\n",
       "      <td>chr7_5555120_5555340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5555056</td>\n",
       "      <td>5556791</td>\n",
       "      <td>chr7:5555056-5556791:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5555880</td>\n",
       "      <td>5556080</td>\n",
       "      <td>1.984910</td>\n",
       "      <td>chr7_5555880_5556080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5555056</td>\n",
       "      <td>5556791</td>\n",
       "      <td>chr7:5555056-5556791:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5556580</td>\n",
       "      <td>5556660</td>\n",
       "      <td>4.553980</td>\n",
       "      <td>chr7_5556580_5556660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5071872</td>\n",
       "      <td>5073109</td>\n",
       "      <td>chr7:5071872-5073109:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5071940</td>\n",
       "      <td>5072180</td>\n",
       "      <td>1.262550</td>\n",
       "      <td>chr7_5071940_5072180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5569296</td>\n",
       "      <td>5570610</td>\n",
       "      <td>chr7:5569296-5570610:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5569460</td>\n",
       "      <td>5569820</td>\n",
       "      <td>1.620590</td>\n",
       "      <td>chr7_5569460_5569820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5569296</td>\n",
       "      <td>5570610</td>\n",
       "      <td>chr7:5569296-5570610:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5570160</td>\n",
       "      <td>5570340</td>\n",
       "      <td>12.135600</td>\n",
       "      <td>chr7_5570160_5570340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134549336</td>\n",
       "      <td>134550882</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134549540</td>\n",
       "      <td>134549700</td>\n",
       "      <td>2.085410</td>\n",
       "      <td>chrX_134549540_134549700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134549336</td>\n",
       "      <td>134550882</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134550780</td>\n",
       "      <td>134551000</td>\n",
       "      <td>1.300240</td>\n",
       "      <td>chrX_134550780_134551000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr1     start1       end1             PerturbTargetID   chr2     start2  \\\n",
       "0    chr2   47191521   47193363    chr2:47191521-47193363:.   chr2   47191548   \n",
       "1    chr2   47191521   47193363    chr2:47191521-47193363:.   chr2   47192620   \n",
       "2   chr20    4760486    4761348     chr20:4760486-4761348:.  chr20    4760680   \n",
       "3   chr20    4760486    4761348     chr20:4760486-4761348:.  chr20    4760960   \n",
       "4    chr7    5486761    5488068      chr7:5486761-5488068:.   chr7    5486960   \n",
       "5    chr7    5486761    5488068      chr7:5486761-5488068:.   chr7    5487920   \n",
       "6    chr7    5607061    5607714      chr7:5607061-5607714:.   chr7    5607560   \n",
       "7    chr7    6525977    6526746      chr7:6525977-6526746:.   chr7    6526560   \n",
       "8    chr7    5426774    5428241      chr7:5426774-5428241:.   chr7    5426800   \n",
       "9    chr7    5426774    5428241      chr7:5426774-5428241:.   chr7    5427360   \n",
       "10   chr7    5426774    5428241      chr7:5426774-5428241:.   chr7    5427860   \n",
       "11   chr7    6663773    6664698      chr7:6663773-6664698:.   chr7    6663840   \n",
       "12   chr7    6081033    6081797      chr7:6081033-6081797:.   chr7    6081460   \n",
       "13   chr7    5555056    5556791      chr7:5555056-5556791:.   chr7    5554920   \n",
       "14   chr7    5555056    5556791      chr7:5555056-5556791:.   chr7    5555120   \n",
       "15   chr7    5555056    5556791      chr7:5555056-5556791:.   chr7    5555880   \n",
       "16   chr7    5555056    5556791      chr7:5555056-5556791:.   chr7    5556580   \n",
       "17   chr7    5071872    5073109      chr7:5071872-5073109:.   chr7    5071940   \n",
       "18   chr7    5569296    5570610      chr7:5569296-5570610:.   chr7    5569460   \n",
       "19   chr7    5569296    5570610      chr7:5569296-5570610:.   chr7    5570160   \n",
       "20   chrX  134549336  134550882  chrX:134549336-134550882:.   chrX  134549540   \n",
       "21   chrX  134549336  134550882  chrX:134549336-134550882:.   chrX  134550780   \n",
       "\n",
       "         end2     signal                    dhs_cc  \n",
       "0    47191727   1.017580    chr2_47191548_47191727  \n",
       "1    47192709   0.866826    chr2_47192620_47192709  \n",
       "2     4760840   2.908270     chr20_4760680_4760840  \n",
       "3     4761280   0.829138     chr20_4760960_4761280  \n",
       "4     5487180   4.026350      chr7_5486960_5487180  \n",
       "5     5488100   2.456010      chr7_5487920_5488100  \n",
       "6     5607760   2.135660      chr7_5607560_5607760  \n",
       "7     6526840   2.116820      chr7_6526560_6526840  \n",
       "8     5426960   2.895700      chr7_5426800_5426960  \n",
       "9     5427634   1.331650      chr7_5427360_5427634  \n",
       "10    5428240   2.964800      chr7_5427860_5428240  \n",
       "11    6664040   3.504990      chr7_6663840_6664040  \n",
       "12    6081680   9.610470      chr7_6081460_6081680  \n",
       "13    5555060   0.841701      chr7_5554920_5555060  \n",
       "14    5555340   1.658280      chr7_5555120_5555340  \n",
       "15    5556080   1.984910      chr7_5555880_5556080  \n",
       "16    5556660   4.553980      chr7_5556580_5556660  \n",
       "17    5072180   1.262550      chr7_5071940_5072180  \n",
       "18    5569820   1.620590      chr7_5569460_5569820  \n",
       "19    5570340  12.135600      chr7_5570160_5570340  \n",
       "20  134549700   2.085410  chrX_134549540_134549700  \n",
       "21  134551000   1.300240  chrX_134550780_134551000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##### (2) Intersect with DHS\n",
    "# Note, the DHS file lacks summit positions!!!\n",
    "WTC11_DHS_file = 'WTC11_DNase_hg38_ENCFF854DSG.bed.gz' # Note, this does not come from the same experiment/line\n",
    "dhs_intersect_ofile = 'NonTSS_Common_Sig_Elements_DHS_Intersect.bed'\n",
    "fopen = open(dhs_intersect_ofile, 'w')\n",
    "subprocess.call(['bedtools','intersect','-a',tssfiltered_common_sig_elements_output,'-b',WTC11_DHS_file,'-wa','-wb'],stdout=fopen)\n",
    "fopen.close()\n",
    "\n",
    "# Process and get DHS peaks\n",
    "dhs_intersect_df = pd.read_csv(dhs_intersect_ofile, sep = '\\t', header = None, names = ['chr1','start1','end1','PerturbTargetID','chr2','start2','end2','null1','null2','null3','signal','null4','null5','null6'], usecols=['chr1','start1','end1','PerturbTargetID','chr2','start2','end2','signal'])\n",
    "dhs_intersect_df['dhs_cc'] = dhs_intersect_df['chr2'].astype(str) + '_' + dhs_intersect_df['start2'].astype(str) + '_' + dhs_intersect_df['end2'].astype(str)\n",
    "\n",
    "# Remove duplicates\n",
    "# Note, there are no duplicates\n",
    "dhs_intersect_df.drop_duplicates(subset = ['PerturbTargetID','dhs_cc'], inplace=True)\n",
    "display(dhs_intersect_df)\n",
    "\n",
    "# Record element_dhs dict\n",
    "element_dhs_dict = dhs_intersect_df.set_index('PerturbTargetID').to_dict()['dhs_cc'] # Key = PerturbTargetID; Value = dhs_cc\n",
    "\n",
    "# Write DHS to output, keep the PerturbTargetID\n",
    "dhs_intersect_ofile2 = 'NonTSS_Common_Sig_Elements_DHS_Intersect_cleaned.bed'\n",
    "dhs_intersect_df.to_csv(dhs_intersect_ofile2, sep = '\\t', columns=['chr2','start2','end2','dhs_cc','PerturbTargetID'], header=False, index=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "ed0f189f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dhs_chr</th>\n",
       "      <th>dhs_start</th>\n",
       "      <th>dhs_end</th>\n",
       "      <th>dhs_cc</th>\n",
       "      <th>PerturbTargetID</th>\n",
       "      <th>h3k27ac_chr</th>\n",
       "      <th>h3k27ac_start</th>\n",
       "      <th>h3k27ac_end</th>\n",
       "      <th>h3k27ac_peak_id</th>\n",
       "      <th>h3k27ac_total_signal</th>\n",
       "      <th>null1</th>\n",
       "      <th>null2</th>\n",
       "      <th>null3</th>\n",
       "      <th>h3k27ac_avg_signal</th>\n",
       "      <th>h3k27ac_summit</th>\n",
       "      <th>h3k27ac_cc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191548</td>\n",
       "      <td>47191727</td>\n",
       "      <td>chr2_47191548_47191727</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191431</td>\n",
       "      <td>47192115</td>\n",
       "      <td>Peak_46664</td>\n",
       "      <td>287</td>\n",
       "      <td>.</td>\n",
       "      <td>5.77803</td>\n",
       "      <td>28.79111</td>\n",
       "      <td>26.25397</td>\n",
       "      <td>588</td>\n",
       "      <td>chr2_47191431_47192115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47192620</td>\n",
       "      <td>47192709</td>\n",
       "      <td>chr2_47192620_47192709</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47192356</td>\n",
       "      <td>47193228</td>\n",
       "      <td>Peak_23593</td>\n",
       "      <td>465</td>\n",
       "      <td>.</td>\n",
       "      <td>4.83537</td>\n",
       "      <td>46.57778</td>\n",
       "      <td>43.75097</td>\n",
       "      <td>407</td>\n",
       "      <td>chr2_47192356_47193228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5607560</td>\n",
       "      <td>5607760</td>\n",
       "      <td>chr7_5607560_5607760</td>\n",
       "      <td>chr7:5607061-5607714:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5607342</td>\n",
       "      <td>5607850</td>\n",
       "      <td>Peak_208671</td>\n",
       "      <td>101</td>\n",
       "      <td>.</td>\n",
       "      <td>3.98150</td>\n",
       "      <td>10.14974</td>\n",
       "      <td>8.17458</td>\n",
       "      <td>103</td>\n",
       "      <td>chr7_5607342_5607850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5427360</td>\n",
       "      <td>5427634</td>\n",
       "      <td>chr7_5427360_5427634</td>\n",
       "      <td>chr7:5426774-5428241:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5427367</td>\n",
       "      <td>5427851</td>\n",
       "      <td>Peak_12084</td>\n",
       "      <td>752</td>\n",
       "      <td>.</td>\n",
       "      <td>13.36647</td>\n",
       "      <td>75.26088</td>\n",
       "      <td>72.10416</td>\n",
       "      <td>157</td>\n",
       "      <td>chr7_5427367_5427851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7</td>\n",
       "      <td>6081460</td>\n",
       "      <td>6081680</td>\n",
       "      <td>chr7_6081460_6081680</td>\n",
       "      <td>chr7:6081033-6081797:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>6081662</td>\n",
       "      <td>6081841</td>\n",
       "      <td>Peak_161857</td>\n",
       "      <td>122</td>\n",
       "      <td>.</td>\n",
       "      <td>4.40809</td>\n",
       "      <td>12.24941</td>\n",
       "      <td>10.17617</td>\n",
       "      <td>93</td>\n",
       "      <td>chr7_6081662_6081841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5555880</td>\n",
       "      <td>5556080</td>\n",
       "      <td>chr7_5555880_5556080</td>\n",
       "      <td>chr7:5555056-5556791:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5555668</td>\n",
       "      <td>5555887</td>\n",
       "      <td>Peak_110371</td>\n",
       "      <td>160</td>\n",
       "      <td>.</td>\n",
       "      <td>5.11907</td>\n",
       "      <td>16.01559</td>\n",
       "      <td>13.80025</td>\n",
       "      <td>124</td>\n",
       "      <td>chr7_5555668_5555887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr7</td>\n",
       "      <td>5569460</td>\n",
       "      <td>5569820</td>\n",
       "      <td>chr7_5569460_5569820</td>\n",
       "      <td>chr7:5569296-5570610:.</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5569258</td>\n",
       "      <td>5569635</td>\n",
       "      <td>Peak_15187</td>\n",
       "      <td>639</td>\n",
       "      <td>.</td>\n",
       "      <td>11.29301</td>\n",
       "      <td>63.96555</td>\n",
       "      <td>60.92708</td>\n",
       "      <td>269</td>\n",
       "      <td>chr7_5569258_5569635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134549540</td>\n",
       "      <td>134549700</td>\n",
       "      <td>chrX_134549540_134549700</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134548604</td>\n",
       "      <td>134549585</td>\n",
       "      <td>Peak_17867</td>\n",
       "      <td>567</td>\n",
       "      <td>.</td>\n",
       "      <td>11.09132</td>\n",
       "      <td>56.76051</td>\n",
       "      <td>53.80297</td>\n",
       "      <td>583</td>\n",
       "      <td>chrX_134548604_134549585</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  dhs_chr  dhs_start    dhs_end                    dhs_cc  \\\n",
       "0    chr2   47191548   47191727    chr2_47191548_47191727   \n",
       "1    chr2   47192620   47192709    chr2_47192620_47192709   \n",
       "2    chr7    5607560    5607760      chr7_5607560_5607760   \n",
       "3    chr7    5427360    5427634      chr7_5427360_5427634   \n",
       "4    chr7    6081460    6081680      chr7_6081460_6081680   \n",
       "5    chr7    5555880    5556080      chr7_5555880_5556080   \n",
       "6    chr7    5569460    5569820      chr7_5569460_5569820   \n",
       "7    chrX  134549540  134549700  chrX_134549540_134549700   \n",
       "\n",
       "              PerturbTargetID h3k27ac_chr  h3k27ac_start  h3k27ac_end  \\\n",
       "0    chr2:47191521-47193363:.        chr2       47191431     47192115   \n",
       "1    chr2:47191521-47193363:.        chr2       47192356     47193228   \n",
       "2      chr7:5607061-5607714:.        chr7        5607342      5607850   \n",
       "3      chr7:5426774-5428241:.        chr7        5427367      5427851   \n",
       "4      chr7:6081033-6081797:.        chr7        6081662      6081841   \n",
       "5      chr7:5555056-5556791:.        chr7        5555668      5555887   \n",
       "6      chr7:5569296-5570610:.        chr7        5569258      5569635   \n",
       "7  chrX:134549336-134550882:.        chrX      134548604    134549585   \n",
       "\n",
       "  h3k27ac_peak_id  h3k27ac_total_signal null1     null2     null3  \\\n",
       "0      Peak_46664                   287     .   5.77803  28.79111   \n",
       "1      Peak_23593                   465     .   4.83537  46.57778   \n",
       "2     Peak_208671                   101     .   3.98150  10.14974   \n",
       "3      Peak_12084                   752     .  13.36647  75.26088   \n",
       "4     Peak_161857                   122     .   4.40809  12.24941   \n",
       "5     Peak_110371                   160     .   5.11907  16.01559   \n",
       "6      Peak_15187                   639     .  11.29301  63.96555   \n",
       "7      Peak_17867                   567     .  11.09132  56.76051   \n",
       "\n",
       "   h3k27ac_avg_signal  h3k27ac_summit                h3k27ac_cc  \n",
       "0            26.25397             588    chr2_47191431_47192115  \n",
       "1            43.75097             407    chr2_47192356_47193228  \n",
       "2             8.17458             103      chr7_5607342_5607850  \n",
       "3            72.10416             157      chr7_5427367_5427851  \n",
       "4            10.17617              93      chr7_6081662_6081841  \n",
       "5            13.80025             124      chr7_5555668_5555887  \n",
       "6            60.92708             269      chr7_5569258_5569635  \n",
       "7            53.80297             583  chrX_134548604_134549585  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##### (3) Intersect with H3K27ac\n",
    "\n",
    "# Intersect\n",
    "H3K27ac_file = 'WTC11_H3K27ac-MintChIPSeq_hg38_ENCFF655PNM.bed.gz'\n",
    "h3k27ac_intersect_ofile = 'NonTSS_Common_Sig_Element_DHS_H3K27ac_intersect.bed'\n",
    "fopen = open(h3k27ac_intersect_ofile, 'w')\n",
    "subprocess.call(['bedtools','intersect','-a',dhs_intersect_ofile2,'-b',H3K27ac_file, '-wa','-wb'],stdout=fopen)\n",
    "fopen.close()\n",
    "\n",
    "# Process intersect\n",
    "h3k27ac_intersect_df = pd.read_csv(h3k27ac_intersect_ofile, sep='\\t', header=None, names=['dhs_chr','dhs_start','dhs_end','dhs_cc','PerturbTargetID','h3k27ac_chr','h3k27ac_start','h3k27ac_end','h3k27ac_peak_id','h3k27ac_total_signal','null1','null2','null3','h3k27ac_avg_signal','h3k27ac_summit'])\n",
    "h3k27ac_intersect_df['h3k27ac_cc'] = h3k27ac_intersect_df['h3k27ac_chr'].astype(str) + '_' + h3k27ac_intersect_df['h3k27ac_start'].astype(str) + '_' + h3k27ac_intersect_df['h3k27ac_end'].astype(str)\n",
    "\n",
    "display(h3k27ac_intersect_df)\n",
    "\n",
    "# Note, the original intersection only provides 8 intersections\n",
    "\n",
    "# Write to output, keep the PerturbTargetID\n",
    "h3k27ac_intersect_ofile2 = 'NonTSS_Common_Sig_Element_DHS_H3K27ac_intersect_cleaned.bed'\n",
    "h3k27ac_intersect_df.to_csv(h3k27ac_intersect_ofile2, sep = '\\t', columns=['h3k27ac_chr','h3k27ac_start','h3k27ac_end','h3k27ac_cc','PerturbTargetID'], index = False, header = False)\n",
    "\n",
    "# ***** MAY COME BACK HERE TO REPEAT THIS AFTER SLOPPING THE DHS PEAKS *****\n",
    "# ***** MAY COME BACK HERE TO REPEAT THIS AFTER SLOPPING THE DHS PEAKS *****\n",
    "# ***** MAY COME BACK HERE TO REPEAT THIS AFTER SLOPPING THE DHS PEAKS *****\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7def514e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nFourth, if an element has multiple DHS peaks that intersect H3K27ac, nominate a single DHS peak. Only one DHS peak per sig element!\\n- Option 1: Use the DHS peak with the highest average (per base) signal.\\n- Option 2: Empirically select the DHS peak closest to the top 3 sgRNAs\\n- Option 3: ALTERNATIVELY, KEEP ALL UNIQUE DHS/H3K27ac PEAKS, AND TREAT THEM INDEPENDENTLY!\\n'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##### (4) Handle elements with multiple unique DHS_CC and H3K27ac peaks\n",
    "# Note, for a first pass, will proceed with ALL peaks, do not try to nominate one DHS/H3K27ac peak per element\n",
    "\"\"\"\n",
    "Fourth, if an element has multiple DHS peaks that intersect H3K27ac, nominate a single DHS peak. Only one DHS peak per sig element!\n",
    "- Option 1: Use the DHS peak with the highest average (per base) signal.\n",
    "- Option 2: Empirically select the DHS peak closest to the top 3 sgRNAs\n",
    "- Option 3: ALTERNATIVELY, KEEP ALL UNIQUE DHS/H3K27ac PEAKS, AND TREAT THEM INDEPENDENTLY!\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "0c1cf7dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>epi_chr</th>\n",
       "      <th>epi_start</th>\n",
       "      <th>epi_end</th>\n",
       "      <th>epi_cc</th>\n",
       "      <th>PerturbTargetID</th>\n",
       "      <th>chrPerturbationTarget</th>\n",
       "      <th>startPerturbationTarget</th>\n",
       "      <th>endPerturbationTarget</th>\n",
       "      <th>name</th>\n",
       "      <th>all_rep_z_log2fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191548</td>\n",
       "      <td>47191727</td>\n",
       "      <td>chr2_47191548_47191727</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191563</td>\n",
       "      <td>47191565</td>\n",
       "      <td>NA|chr2:47191563-47191565:+</td>\n",
       "      <td>0.189141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191548</td>\n",
       "      <td>47191727</td>\n",
       "      <td>chr2_47191548_47191727</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191565</td>\n",
       "      <td>47191567</td>\n",
       "      <td>NA|chr2:47191565-47191567:-</td>\n",
       "      <td>-0.374988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191548</td>\n",
       "      <td>47191727</td>\n",
       "      <td>chr2_47191548_47191727</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191602</td>\n",
       "      <td>47191604</td>\n",
       "      <td>NA|chr2:47191602-47191604:+</td>\n",
       "      <td>-0.564378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191548</td>\n",
       "      <td>47191727</td>\n",
       "      <td>chr2_47191548_47191727</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191620</td>\n",
       "      <td>47191622</td>\n",
       "      <td>NA|chr2:47191620-47191622:+</td>\n",
       "      <td>-0.197447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191548</td>\n",
       "      <td>47191727</td>\n",
       "      <td>chr2_47191548_47191727</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191630</td>\n",
       "      <td>47191632</td>\n",
       "      <td>NA|chr2:47191630-47191632:+</td>\n",
       "      <td>1.647936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134550780</td>\n",
       "      <td>134551000</td>\n",
       "      <td>chrX_134550780_134551000</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134550880</td>\n",
       "      <td>134550882</td>\n",
       "      <td>NA|chrX:134550880-134550882:+</td>\n",
       "      <td>-0.612369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134550780</td>\n",
       "      <td>134551000</td>\n",
       "      <td>chrX_134550780_134551000</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134550858</td>\n",
       "      <td>134550860</td>\n",
       "      <td>NA|chrX:134550858-134550860:-</td>\n",
       "      <td>-0.380782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134550780</td>\n",
       "      <td>134551000</td>\n",
       "      <td>chrX_134550780_134551000</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134550895</td>\n",
       "      <td>134550897</td>\n",
       "      <td>NA|chrX:134550895-134550897:+</td>\n",
       "      <td>-0.121992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134550780</td>\n",
       "      <td>134551000</td>\n",
       "      <td>chrX_134550780_134551000</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134550898</td>\n",
       "      <td>134550900</td>\n",
       "      <td>NA|chrX:134550898-134550900:+</td>\n",
       "      <td>-0.344531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134550780</td>\n",
       "      <td>134551000</td>\n",
       "      <td>chrX_134550780_134551000</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134550877</td>\n",
       "      <td>134550879</td>\n",
       "      <td>NA|chrX:134550877-134550879:-</td>\n",
       "      <td>-0.399807</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>427 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    epi_chr  epi_start    epi_end                    epi_cc  \\\n",
       "0      chr2   47191548   47191727    chr2_47191548_47191727   \n",
       "1      chr2   47191548   47191727    chr2_47191548_47191727   \n",
       "2      chr2   47191548   47191727    chr2_47191548_47191727   \n",
       "3      chr2   47191548   47191727    chr2_47191548_47191727   \n",
       "4      chr2   47191548   47191727    chr2_47191548_47191727   \n",
       "..      ...        ...        ...                       ...   \n",
       "422    chrX  134550780  134551000  chrX_134550780_134551000   \n",
       "423    chrX  134550780  134551000  chrX_134550780_134551000   \n",
       "424    chrX  134550780  134551000  chrX_134550780_134551000   \n",
       "425    chrX  134550780  134551000  chrX_134550780_134551000   \n",
       "426    chrX  134550780  134551000  chrX_134550780_134551000   \n",
       "\n",
       "                PerturbTargetID chrPerturbationTarget  \\\n",
       "0      chr2:47191521-47193363:.                  chr2   \n",
       "1      chr2:47191521-47193363:.                  chr2   \n",
       "2      chr2:47191521-47193363:.                  chr2   \n",
       "3      chr2:47191521-47193363:.                  chr2   \n",
       "4      chr2:47191521-47193363:.                  chr2   \n",
       "..                          ...                   ...   \n",
       "422  chrX:134549336-134550882:.                  chrX   \n",
       "423  chrX:134549336-134550882:.                  chrX   \n",
       "424  chrX:134549336-134550882:.                  chrX   \n",
       "425  chrX:134549336-134550882:.                  chrX   \n",
       "426  chrX:134549336-134550882:.                  chrX   \n",
       "\n",
       "     startPerturbationTarget  endPerturbationTarget  \\\n",
       "0                   47191563               47191565   \n",
       "1                   47191565               47191567   \n",
       "2                   47191602               47191604   \n",
       "3                   47191620               47191622   \n",
       "4                   47191630               47191632   \n",
       "..                       ...                    ...   \n",
       "422                134550880              134550882   \n",
       "423                134550858              134550860   \n",
       "424                134550895              134550897   \n",
       "425                134550898              134550900   \n",
       "426                134550877              134550879   \n",
       "\n",
       "                              name  all_rep_z_log2fc  \n",
       "0      NA|chr2:47191563-47191565:+          0.189141  \n",
       "1      NA|chr2:47191565-47191567:-         -0.374988  \n",
       "2      NA|chr2:47191602-47191604:+         -0.564378  \n",
       "3      NA|chr2:47191620-47191622:+         -0.197447  \n",
       "4      NA|chr2:47191630-47191632:+          1.647936  \n",
       "..                             ...               ...  \n",
       "422  NA|chrX:134550880-134550882:+         -0.612369  \n",
       "423  NA|chrX:134550858-134550860:-         -0.380782  \n",
       "424  NA|chrX:134550895-134550897:+         -0.121992  \n",
       "425  NA|chrX:134550898-134550900:+         -0.344531  \n",
       "426  NA|chrX:134550877-134550879:-         -0.399807  \n",
       "\n",
       "[427 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "epi_cc\n",
       "chr20_4760680_4760840       0.143323\n",
       "chr20_4760960_4761280      -0.148852\n",
       "chr2_47191548_47191727      0.279352\n",
       "chr2_47192620_47192709     -0.191931\n",
       "chr7_5071940_5072180       -0.021583\n",
       "chr7_5426800_5426960       -0.023514\n",
       "chr7_5427360_5427634       -0.045770\n",
       "chr7_5427860_5428240       -0.060878\n",
       "chr7_5486960_5487180       -0.221552\n",
       "chr7_5487920_5488100        0.007402\n",
       "chr7_5554920_5555060       -1.038194\n",
       "chr7_5555120_5555340       -0.249265\n",
       "chr7_5555880_5556080        0.256044\n",
       "chr7_5556580_5556660       -0.111073\n",
       "chr7_5569460_5569820       -0.176050\n",
       "chr7_5570160_5570340        0.014211\n",
       "chr7_5607560_5607760        0.012159\n",
       "chr7_6081460_6081680        0.085489\n",
       "chr7_6526560_6526840       -0.038710\n",
       "chr7_6663840_6664040       -0.303666\n",
       "chrX_134549540_134549700   -0.171844\n",
       "chrX_134550780_134551000   -0.367152\n",
       "Name: all_rep_z_log2fc, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>epi_chr</th>\n",
       "      <th>epi_start</th>\n",
       "      <th>epi_end</th>\n",
       "      <th>epi_cc</th>\n",
       "      <th>PerturbTargetID</th>\n",
       "      <th>chrPerturbationTarget</th>\n",
       "      <th>startPerturbationTarget</th>\n",
       "      <th>endPerturbationTarget</th>\n",
       "      <th>name</th>\n",
       "      <th>all_rep_z_log2fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191431</td>\n",
       "      <td>47192115</td>\n",
       "      <td>chr2_47191431_47192115</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191530</td>\n",
       "      <td>47191532</td>\n",
       "      <td>NA|chr2:47191530-47191532:+</td>\n",
       "      <td>-0.870077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191431</td>\n",
       "      <td>47192115</td>\n",
       "      <td>chr2_47191431_47192115</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191521</td>\n",
       "      <td>47191523</td>\n",
       "      <td>NA|chr2:47191521-47191523:-</td>\n",
       "      <td>-0.381478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191431</td>\n",
       "      <td>47192115</td>\n",
       "      <td>chr2_47191431_47192115</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191522</td>\n",
       "      <td>47191524</td>\n",
       "      <td>NA|chr2:47191522-47191524:-</td>\n",
       "      <td>-0.484247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191431</td>\n",
       "      <td>47192115</td>\n",
       "      <td>chr2_47191431_47192115</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191563</td>\n",
       "      <td>47191565</td>\n",
       "      <td>NA|chr2:47191563-47191565:+</td>\n",
       "      <td>0.189141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>47191431</td>\n",
       "      <td>47192115</td>\n",
       "      <td>chr2_47191431_47192115</td>\n",
       "      <td>chr2:47191521-47193363:.</td>\n",
       "      <td>chr2</td>\n",
       "      <td>47191565</td>\n",
       "      <td>47191567</td>\n",
       "      <td>NA|chr2:47191565-47191567:-</td>\n",
       "      <td>-0.374988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134548604</td>\n",
       "      <td>134549585</td>\n",
       "      <td>chrX_134548604_134549585</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134549536</td>\n",
       "      <td>134549538</td>\n",
       "      <td>NA|chrX:134549536-134549538:+</td>\n",
       "      <td>0.726574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134548604</td>\n",
       "      <td>134549585</td>\n",
       "      <td>chrX_134548604_134549585</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134549560</td>\n",
       "      <td>134549562</td>\n",
       "      <td>NA|chrX:134549560-134549562:+</td>\n",
       "      <td>1.150136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134548604</td>\n",
       "      <td>134549585</td>\n",
       "      <td>chrX_134548604_134549585</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134549567</td>\n",
       "      <td>134549569</td>\n",
       "      <td>NA|chrX:134549567-134549569:+</td>\n",
       "      <td>-0.293044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134548604</td>\n",
       "      <td>134549585</td>\n",
       "      <td>chrX_134548604_134549585</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134549575</td>\n",
       "      <td>134549577</td>\n",
       "      <td>NA|chrX:134549575-134549577:+</td>\n",
       "      <td>-0.124921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>chrX</td>\n",
       "      <td>134548604</td>\n",
       "      <td>134549585</td>\n",
       "      <td>chrX_134548604_134549585</td>\n",
       "      <td>chrX:134549336-134550882:.</td>\n",
       "      <td>chrX</td>\n",
       "      <td>134548711</td>\n",
       "      <td>134548713</td>\n",
       "      <td>NA|chrX:134548711-134548713:+</td>\n",
       "      <td>0.106586</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>232 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    epi_chr  epi_start    epi_end                    epi_cc  \\\n",
       "0      chr2   47191431   47192115    chr2_47191431_47192115   \n",
       "1      chr2   47191431   47192115    chr2_47191431_47192115   \n",
       "2      chr2   47191431   47192115    chr2_47191431_47192115   \n",
       "3      chr2   47191431   47192115    chr2_47191431_47192115   \n",
       "4      chr2   47191431   47192115    chr2_47191431_47192115   \n",
       "..      ...        ...        ...                       ...   \n",
       "227    chrX  134548604  134549585  chrX_134548604_134549585   \n",
       "228    chrX  134548604  134549585  chrX_134548604_134549585   \n",
       "229    chrX  134548604  134549585  chrX_134548604_134549585   \n",
       "230    chrX  134548604  134549585  chrX_134548604_134549585   \n",
       "231    chrX  134548604  134549585  chrX_134548604_134549585   \n",
       "\n",
       "                PerturbTargetID chrPerturbationTarget  \\\n",
       "0      chr2:47191521-47193363:.                  chr2   \n",
       "1      chr2:47191521-47193363:.                  chr2   \n",
       "2      chr2:47191521-47193363:.                  chr2   \n",
       "3      chr2:47191521-47193363:.                  chr2   \n",
       "4      chr2:47191521-47193363:.                  chr2   \n",
       "..                          ...                   ...   \n",
       "227  chrX:134549336-134550882:.                  chrX   \n",
       "228  chrX:134549336-134550882:.                  chrX   \n",
       "229  chrX:134549336-134550882:.                  chrX   \n",
       "230  chrX:134549336-134550882:.                  chrX   \n",
       "231  chrX:134549336-134550882:.                  chrX   \n",
       "\n",
       "     startPerturbationTarget  endPerturbationTarget  \\\n",
       "0                   47191530               47191532   \n",
       "1                   47191521               47191523   \n",
       "2                   47191522               47191524   \n",
       "3                   47191563               47191565   \n",
       "4                   47191565               47191567   \n",
       "..                       ...                    ...   \n",
       "227                134549536              134549538   \n",
       "228                134549560              134549562   \n",
       "229                134549567              134549569   \n",
       "230                134549575              134549577   \n",
       "231                134548711              134548713   \n",
       "\n",
       "                              name  all_rep_z_log2fc  \n",
       "0      NA|chr2:47191530-47191532:+         -0.870077  \n",
       "1      NA|chr2:47191521-47191523:-         -0.381478  \n",
       "2      NA|chr2:47191522-47191524:-         -0.484247  \n",
       "3      NA|chr2:47191563-47191565:+          0.189141  \n",
       "4      NA|chr2:47191565-47191567:-         -0.374988  \n",
       "..                             ...               ...  \n",
       "227  NA|chrX:134549536-134549538:+          0.726574  \n",
       "228  NA|chrX:134549560-134549562:+          1.150136  \n",
       "229  NA|chrX:134549567-134549569:+         -0.293044  \n",
       "230  NA|chrX:134549575-134549577:+         -0.124921  \n",
       "231  NA|chrX:134548711-134548713:+          0.106586  \n",
       "\n",
       "[232 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "epi_cc\n",
       "chr2_47191431_47192115      0.053818\n",
       "chr2_47192356_47193228     -0.099076\n",
       "chr7_5427367_5427851       -0.027801\n",
       "chr7_5555668_5555887       -0.070976\n",
       "chr7_5569258_5569635        0.019631\n",
       "chr7_5607342_5607850        0.056293\n",
       "chr7_6081662_6081841        0.074098\n",
       "chrX_134548604_134549585   -0.058290\n",
       "Name: all_rep_z_log2fc, dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##### (5) Get avg effects of all sgRNAs intersecting a given epifeature peak\n",
    "\"\"\"\n",
    "Fifth, get the average effects of all sgRNAs intersecting a given epifeature peak.\n",
    "- Effect is the average z-transformed log2FC from both screen replicates\n",
    "- sgRNA intersections are based on PAM coordinates, with no further slop/range expansion\n",
    "- If a DHS intersects two H3K27ac peaks, treat each H3K27ac peak independently\n",
    "\"\"\"\n",
    "# Input files\n",
    "dhs_element_file = dhs_intersect_ofile2\n",
    "h3k27ac_element_file = h3k27ac_intersect_ofile2\n",
    "grna_pam_file = 'WTC11_CRISPRi_{0}_{1}_Processed_Simple_PamCC.bed'.format(treatment_tag,control_tag)\n",
    "\n",
    "## Intersect epifiles with grna file\n",
    "# DHS - guide\n",
    "dhs_grna_intersect_file = 'dhs_grna_intersect.bed'\n",
    "fopen = open(dhs_grna_intersect_file, 'w')\n",
    "subprocess.call(['bedtools','intersect','-a',dhs_element_file,'-b',grna_pam_file,'-wa','-wb'],stdout=fopen)\n",
    "\n",
    "fopen.close()\n",
    "# H3K27ac - guide\n",
    "h3k27ac_grna_intersect_file = 'h3k27ac_grna_intersect.bed'\n",
    "fopen = open(h3k27ac_grna_intersect_file, 'w')\n",
    "subprocess.call(['bedtools','intersect','-a',h3k27ac_element_file,'-b',grna_pam_file,'-wa','-wb'],stdout=fopen)\n",
    "fopen.close()\n",
    "\n",
    "### Process intersections\n",
    "for f in [dhs_grna_intersect_file, h3k27ac_grna_intersect_file]:\n",
    "    # Read file to df\n",
    "    epifeature_grna_intersect_df = pd.read_csv(f, sep = '\\t', header=None, names = ['epi_chr','epi_start','epi_end','epi_cc','PerturbTargetID','chrPerturbationTarget','startPerturbationTarget','endPerturbationTarget','name','all_rep_z_log2fc'])\n",
    "    display(epifeature_grna_intersect_df)\n",
    "    # For each epi_cc, use groupby to get average all_rep_z_log2fc\n",
    "    averaged_df = epifeature_grna_intersect_df.groupby('epi_cc')['all_rep_z_log2fc'].mean()\n",
    "    \n",
    "    display(averaged_df)\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
